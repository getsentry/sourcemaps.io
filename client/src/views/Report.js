import React, { Component } from 'react';
import ReactTooltip from 'react-tooltip';
import PropTypes from 'prop-types';

import Loader from './Loader';

const STORAGE_URL = process.env.REACT_APP_STORAGE_URL;

/**
 * Error entry for token/mapping errors (includes source code context)
 */
function BadTokenEntry(props, index) {
  const { source, expected, token, mapping } = props;
  const {
    originalContext,
    originalLine,
    originalColumn,
    generatedColumn,
    generatedLine
  } = mapping;
  return (
    <li className="entry" key={index}>
      <ReactTooltip id={`tt_${index}`} effect="solid">
        Generated location: L{generatedLine}
        {':'}
        {generatedColumn}
      </ReactTooltip>
      <p>
        In <code>{source}</code>
        {':'} Expected <code>{expected}</code> but got <code>{token}</code> at{' '}
        <span data-tip data-for={`tt_${index}`}>
          {`L${originalLine}:${originalColumn}`}
        </span>
      </p>
      <div>
        <pre className="code">
          <ol start={originalContext[0][0]}>
            {originalContext.map(([line, ctx]) => (
              <li
                className={line === originalLine ? 'line line-active' : 'line'}
                key={line}
              >
                <span>{ctx}</span>
              </li>
            ))}
          </ol>
        </pre>
      </div>
    </li>
  );
}

BadTokenEntry.propTypes = {
  source: PropTypes.string.isRequired,
  expected: PropTypes.string.isRequired,
  token: PropTypes.string.isRequired,
  mapping: PropTypes.object.isRequired
};

/**
 * Generic error entry
 */
function Entry(props, index) {
  const { name, message, resolutions } = props;

  const htmlMessage = (message || '').replace(
    /(https?\S+)/g,
    '<a href="$1">$1</a>'
  );
  return (
    <li className="entry" key={index}>
      <h4>{name}</h4>
      {message && <p className="rr-block" dangerouslySetInnerHTML={{ __html: htmlMessage }} />}
      {resolutions && resolutions.length > 0 && (
        <div>
          <h5>Resolutions</h5>
          <ul>
            {resolutions.map((res, i) => (
              <li key={i} dangerouslySetInnerHTML={{ __html: res }} />
            ))}
          </ul>
        </div>
      )}
    </li>
  );
}

Entry.propTypes = {
  name: PropTypes.string.isRequired,
  message: PropTypes.string.isRequired,
  resolutions: PropTypes.array,
  source: PropTypes.string,
  column: PropTypes.number,
  line: PropTypes.number,
  expected: PropTypes.string,
  token: PropTypes.string
};

class Report extends Component {
  constructor(props) {
    super(props);
    const reportName = props.match.params.report;
    this.state = { reportName, report: null };
  }

  componentDidMount() {
    const { reportName } = this.state;
    const reportUrl = `${STORAGE_URL}/${encodeURIComponent(reportName)}`;

    fetch(reportUrl).then((response) => {
      response.json().then((report) => {
        this.setState({ report });
      });
    });
  }

  renderAlert() {
    const { report } = this.state;
    if (!report) return null;

    if (report.errors.length === 0 && report.warnings.length === 0) {
      return (
        <div className="alert alert-success">
          <strong>Bingo.</strong> Everything looks good.
        </div>
      );
    } else if (report.errors.length > 0) {
      return (
        <div className="alert alert-danger">
          <strong>Ouch.</strong> Check the errors below.
        </div>
      );
    }
    return (
      <div className="alert alert-warning">
        <strong>Hmm.</strong> Successful but with warnings.
      </div>
    );
  }

  render() {
    const { report } = this.state;
    return !report ? (
      <Loader />
    ) : (
      <div>
        {this.renderAlert()}
        <h2>
          Report for <a href={report.url} className="rr-mask">{report.url.split('/').pop()}</a>
        </h2>
        {report && (
          <div>
            <h3>URL</h3>
            <ul>
              <li>
                <a href={report.url} className="rr-mask">{report.url}</a>
              </li>
            </ul>
            {report.sourceMap && (
              <div>
                <h3>Source Map</h3>
                <ul>
                  <li>
                    <a href={report.sourceMap} className="rr-mask">{report.sourceMap}</a>
                  </li>
                </ul>
              </div>
            )}
            {report.sources.length > 0 && (
              <div>
                <h3>
                  Sources{' '}
                  <span className="badge badge-success">
                    {report.sources.length}
                  </span>
                </h3>
                <ul>
                  {report.sources.map(src => (
                    <li key={src}>
                      <a href={src} className="rr-mask">{src}</a>
                    </li>
                  ))}
                </ul>
              </div>
            )}
            <h3>
              Errors{' '}
              {report.errors.length > 0 ? (
                <span className="badge" style={{ background: '#b94a48' }}>
                  {report.errors.length}
                </span>
              ) : (
                <span className="badge">0</span>
              )}
            </h3>
            <ul>
              {report.errors.map((err, index) => {
                return ['BadTokenError', 'BadColumnError'].includes(err.name)
                  ? BadTokenEntry(err, index)
                  : Entry(err, index);
              })}
            </ul>
            <h3>
              Warnings{' '}
              {report.warnings.length > 0 ? (
                <span className="badge" style={{ background: '#f89406' }}>
                  {report.warnings.length}
                </span>
              ) : (
                <span className="badge">0</span>
              )}
            </h3>
            <ul>
              {report.warnings.map((err, index) => {
                return ['BadTokenError', 'BadColumnError'].includes(err.name)
                  ? BadTokenEntry(err, index)
                  : Entry(err, index);
              })}
            </ul>
          </div>
        )}
      </div>
    );
  }
}

Report.propTypes = {
  location: PropTypes.object,
  match: PropTypes.object
};

export default Report;
