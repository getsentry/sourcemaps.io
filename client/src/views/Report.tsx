import {Tooltip as ReactTooltip} from 'react-tooltip';

import Loader from './Loader';
import { useEffect, useState } from 'react';
import { useParams } from 'react-router-dom';

const STORAGE_URL = import.meta.env.REACT_APP_STORAGE_URL;

type LineOrColumn = number;
interface ErrorType {
  name: string,
  message: string,
  resolutions?: string[],
  source?: string,
  column?: number,
  line?: number,
  expected?: string,
  token?: string,
  mapping: {
    originalContext: [LineOrColumn, string][],
    originalLine: LineOrColumn,
    originalColumn: LineOrColumn,
    generatedColumn: LineOrColumn,
    generatedLine: LineOrColumn,
  },
}

interface Report {
  url: string;
  errors: ErrorType[];
  warnings: ErrorType[];
  sources: string[]
  sourceMap?: string;
}


interface EntryProps {
  error: ErrorType,
  index: number
};

/**
 * Error entry for token/mapping errors (includes source code context)
 */
function BadTokenEntry({error: {source, expected, token, mapping}, index}: EntryProps) {
  const {
    originalContext,
    originalLine,
    originalColumn,
    generatedColumn,
    generatedLine
  } = mapping;
  return (
    <li className="entry" key={index}>
      <ReactTooltip id={`tt_${index}`}>
        Generated location: L{generatedLine}
        {':'}
        {generatedColumn}
      </ReactTooltip>
      <p>
        In <code>{source}</code>
        {':'} Expected <code>{expected}</code> but got <code>{token}</code> at{' '}
        <span data-tip data-for={`tt_${index}`}>
          {`L${originalLine}:${originalColumn}`}
        </span>
      </p>
      <div>
        <pre className="code">
          <ol start={originalContext[0][0]}>
            {originalContext.map(([line, ctx]) => (
              <li
                className={line === originalLine ? 'line line-active' : 'line'}
                key={line}
              >
                <span>{ctx}</span>
              </li>
            ))}
          </ol>
        </pre>
      </div>
    </li>
  );
}

/**
 * Generic error entry
 */
function Entry({error, index}: EntryProps) {
  const { name, message, resolutions } = error;

  const htmlMessage = (message || '').replace(
    /(https?\S+)/g,
    '<a href="$1">$1</a>'
  );
  return (
    <li className="entry" key={index}>
      <h4>{name}</h4>
      {message && <p className="sr-block" dangerouslySetInnerHTML={{ __html: htmlMessage }} />}
      {resolutions && resolutions.length > 0 && (
        <div>
          <h5>Resolutions</h5>
          <ul>
            {resolutions.map((res, i) => (
              <li key={i} dangerouslySetInnerHTML={{ __html: res }} />
            ))}
          </ul>
        </div>
      )}
    </li>
  );
}


function Alert({report}: {report: Report}) {
  if (!report) {
    return null;
  }

  if (report.errors.length === 0 && report.warnings.length === 0) {
    return (
      <div className="alert alert-success">
        <strong>Bingo.</strong> Everything looks good.
      </div>
    );
  } if (report.errors.length > 0) {
    return (
      <div className="alert alert-danger">
        <strong>Ouch.</strong> Check the errors below.
      </div>
    );
  }
  return (
    <div className="alert alert-warning">
      <strong>Hmm.</strong> Successful but with warnings.
    </div>
  );
}


// Report.propTypes = {
//   location: PropTypes.object,
//   match: PropTypes.object
// };

export default function Report() {
  const {reportName} = useParams();
  const [report, setReport] = useState<Report | null>(null);

  useEffect(() => {
    const reportUrl = `${STORAGE_URL}/${encodeURIComponent(reportName ?? '')}`;
    fetch(reportUrl).then(response => response.json()).then(setReport);
  }, [report, reportName]);
  
  if (!report) {
    return <Loader />;
  };

  return (
    <div>
      <Alert report={report} />
      <h2>
        Report for <a href={report.url} className="sr-mask">{report.url.split('/').pop()}</a>
      </h2>
      <div>
        <h3>URL</h3>
        <ul>
          <li>
            <a href={report.url} className="sr-mask">{report.url}</a>
          </li>
        </ul>
        {report.sourceMap && (
          <div>
            <h3>Source Map</h3>
            <ul>
              <li>
                <a href={report.sourceMap} className="sr-mask">{report.sourceMap}</a>
              </li>
            </ul>
          </div>
        )}
        {report.sources.length > 0 && (
          <div>
            <h3>
              Sources{' '}
              <span className="badge badge-success">
                {report.sources.length}
              </span>
            </h3>
            <ul>
              {report.sources.map(src => (
                <li key={src}>
                  <a href={src} className="sr-mask">{src}</a>
                </li>
              ))}
            </ul>
          </div>
        )}
        <h3>
          Errors{' '}
          {report.errors.length > 0 ? (
            <span className="badge" style={{ background: '#b94a48' }}>
              {report.errors.length}
            </span>
          ) : (
            <span className="badge">0</span>
          )}
        </h3>
        <ul>
          {report.errors.map((error, index) => {
            return ['BadTokenError', 'BadColumnError'].includes(error.name)
              ? <BadTokenEntry error={error} index={index} />
              : <Entry error={error} index={index} />;
          })}
        </ul>
        <h3>
          Warnings{' '}
          {report.warnings.length > 0 ? (
            <span className="badge" style={{ background: '#f89406' }}>
              {report.warnings.length}
            </span>
          ) : (
            <span className="badge">0</span>
          )}
        </h3>
        <ul>
          {report.warnings.map((error, index) => {
            return ['BadTokenError', 'BadColumnError'].includes(error.name)
              ? <BadTokenEntry error={error} index={index} />
              : <Entry error={error} index={index} />;
          })}
        </ul>
      </div>
    </div>
  );
}
