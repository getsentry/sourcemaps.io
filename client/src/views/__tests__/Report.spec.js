import React from 'react';
import {render, waitFor, screen, waitForElementToBeRemoved} from '@testing-library/react';
import fetchMock from 'fetch-mock';

import Report from '../Report';

const exampleReport = {
  errors: [],
  warnings: [],
  sources: [],
  sourceMap: '',
  url: ''
};

it('should fetch the report URL corresponding to the route param', (done) => {
  fetchMock.get('undefined/report.json', () => {
    fetchMock.restore();
    return JSON.stringify(exampleReport);
  });

  const {container} = render(<Report match={{params: { report: 'report.json' }}}/>);
  waitFor(() => screen.getByTestId('loader')).then(() => {
    waitForElementToBeRemoved(() => screen.getByTestId('loader')).then(() => {
      expect(container.firstChild).toMatchSnapshot();
      done();
    });
  });
});
