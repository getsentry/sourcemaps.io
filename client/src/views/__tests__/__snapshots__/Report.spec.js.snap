// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should fetch the report URL corresponding to the route param 1`] = `
<div
  data-reactroot=""
>
  <div
    class="alert alert-success"
  >
    <strong>
      Bingo.
    </strong>
    <!-- react-text: 15 -->
     Everything looks good.
    <!-- /react-text -->
  </div>
  <h2>
    <!-- react-text: 17 -->
    Report for 
    <!-- /react-text -->
    <a
      class="rr-mask"
      href=""
    />
  </h2>
  <div>
    <h3>
      URL
    </h3>
    <ul>
      <li>
        <a
          class="rr-mask"
          href=""
        />
      </li>
    </ul>
    <!-- react-text: 24 -->
    <!-- /react-text -->
    <h3>
      <!-- react-text: 26 -->
      Errors
      <!-- /react-text -->
      <!-- react-text: 27 -->
       
      <!-- /react-text -->
      <span
        class="badge"
      >
        0
      </span>
    </h3>
    <ul />
    <h3>
      <!-- react-text: 31 -->
      Warnings
      <!-- /react-text -->
      <!-- react-text: 32 -->
       
      <!-- /react-text -->
      <span
        class="badge"
      >
        0
      </span>
    </h3>
    <ul />
  </div>
</div>
`;
